File: ch31/App_Code/CS/Listing31_8.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net.Http;
using System.Threading.Tasks;
using System.Web;

public class Listing31_8 : IHttpModule
{
    public void Init(HttpApplication context)
    {
        EventHandlerTaskAsyncHelper helper =
           new EventHandlerTaskAsyncHelper(DownloadWeb);
        
        context.AddOnBeginRequestAsync(
                helper.BeginEventHandler, helper.EndEventHandler);
    }
    public async Task DownloadWeb(object caller, EventArgs e)
    {
        using (var client = new HttpClient())
        {
            var result = await client.GetStringAsync("http://bing.com");            
        }
    }

    public void Dispose()
    {

    }
}

File: ch31/App_Code/VB/Listing31_8.vb
Imports Microsoft.VisualBasic
Imports System.Net.Http
Imports System.Threading.Tasks

Public Class Listing31_8
    Implements IHttpModule
    Public Sub Init(context As HttpApplication) Implements IHttpModule.Init
        Dim helper As New EventHandlerTaskAsyncHelper(AddressOf DownloadWeb)
        context.AddOnBeginRequestAsync(helper.BeginEventHandler, helper.EndEventHandler)
    End Sub
    Public Async Function DownloadWeb(caller As Object, e As EventArgs) As Task
        Using client = New HttpClient()
            Dim result = Await client.GetStringAsync("http://bing.com")
        End Using
    End Function

    Public Sub Dispose() Implements IHttpModule.Dispose

    End Sub
End Class
